ADD pod resource & routes to get member-cluster namespace/deployment/pod

Heylosky opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add pod resources. Add routes for obtaining member cluster pods( list and detail ), namespaces, and deployments.
Set a router group under ``/api/v1`` for prefix ``/member/:clustername``, to get member cluster resources.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

warjiang commented:
/assign

warjiang commented:
Thansk, i'll check it today~

warjiang commented on cmd/api/app/routes/member/deployment/handler.go:
```diff
@@ -0,0 +1,76 @@
+	if err != nil {
+		common.Fail(c, err)
+		return
+	}
```
duplicate check for existence of member cluster, for these kind of check, we can use middleware in gin, like :
```
// middleware.go
func EnsureMemberClusterMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		karmadaClient := client.InClusterKarmadaClient()
		_, err := karmadaClient.ClusterV1alpha1().Clusters().Get(context.TODO(), c.Param("clustername"), metav1.GetOptions{})
		if err != nil {
			c.AbortWithStatusJSON(http.StatusOK, common.BaseResponse{
				Code: 500,
				Msg:  err.Error(),
			})
			return
		}
		c.Next()
	}
}

// usage in setup.go
member = v1.Group("/member/:clustername")
member.Use(EnsureMemberClusterMiddleware())
```
when the code executing in the handler, we can ensure the existence of membercluster, otherwise it will be abort in the middleware.

warjiang commented:
I've check the code locally, the only one problem is duplicate check of membercluster. Apart from that, one question I want to talk about is that for all kind of resources in member cluster, is it need to make another copy just like the ones we implemented for the karmada apiserver. It seems that the answer is yes, but is there a better to implemented that. Just a question, i have no idea till now.

Heylosky commented:
> I've check the code locally, the only one problem is duplicate check of membercluster. Apart from that, one question I want to talk about is that for all kind of resources in member cluster, is it need to make another copy just like the ones we implemented for the karmada apiserver. It seems that the answer is yes, but is there a better to implemented that. Just a question, i have no idea till now.

Perhaps add a middleware for the karmada apiserver as well, and put the client initialization process in the middleware, and then pass the client using the `c.Set` and `c.Get`? This way we can use the same implementation across all resources in the member cluster and the karmada control plane. Just like:
```
func EnsureMemberClusterMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		karmadaClient := client.InClusterKarmadaClient()
		_, err := karmadaClient.ClusterV1alpha1().Clusters().Get(context.TODO(), c.Param("clustername"), metav1.GetOptions{})
		if err != nil {
			c.AbortWithStatusJSON(http.StatusOK, common.BaseResponse{
				Code: 500,
				Msg:  err.Error(),
			})
			return
		}
		// Init client for member cluster
		memberClient := client.InClusterClientForMemberCluster(c.Param("clustername"))
		c.Set("client", memberClient)

		c.Next()
	}
}

func V1Middleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		// Init client for controller cluster
		k8sClient := client.InClusterClientForKarmadaApiServer()
		c.Set("client", k8sClient)

		c.Next()
	}
}
```
For example, to get deployments:
```
func handleGetDeployments(c *gin.Context) {
	if value, exists := c.Get("client"); exists {
		if memberClient, ok := value.(kubeclient.Interface); ok {
			namespace := common.ParseNamespacePathParameter(c)
			dataSelect := common.ParseDataSelectPathParameter(c)
			result, err := deployment.GetDeploymentList(memberClient, namespace, dataSelect)
			if err != nil {
				common.Fail(c, err)
				return
			}
			common.Success(c, result)
		} else {
			c.JSON(500, gin.H{"error": "Failed to assert memberClient to kubernetes.Interface"})
		}
	} else {
		c.JSON(500, gin.H{"error": "memberClient not found"})
	}
}
```
In this case, maybe we can avoid some copy?

warjiang commented:
> > I've check the code locally, the only one problem is duplicate check of membercluster. Apart from that, one question I want to talk about is that for all kind of resources in member cluster, is it need to make another copy just like the ones we implemented for the karmada apiserver. It seems that the answer is yes, but is there a better to implemented that. Just a question, i have no idea till now.
> 
> Perhaps add a middleware for the karmada apiserver as well, and put the client initialization process in the middleware, and then pass the client using the `c.Set` and `c.Get`? This way we can use the same implementation across all resources in the member cluster and the karmada control plane. Just like:
> 
> ```
> func EnsureMemberClusterMiddleware() gin.HandlerFunc {
> 	return func(c *gin.Context) {
> 		karmadaClient := client.InClusterKarmadaClient()
> 		_, err := karmadaClient.ClusterV1alpha1().Clusters().Get(context.TODO(), c.Param("clustername"), metav1.GetOptions{})
> 		if err != nil {
> 			c.AbortWithStatusJSON(http.StatusOK, common.BaseResponse{
> 				Code: 500,
> 				Msg:  err.Error(),
> 			})
> 			return
> 		}
> 		// Init client for member cluster
> 		memberClient := client.InClusterClientForMemberCluster(c.Param("clustername"))
> 		c.Set("client", memberClient)
> 
> 		c.Next()
> 	}
> }
> 
> func V1Middleware() gin.HandlerFunc {
> 	return func(c *gin.Context) {
> 		// Init client for controller cluster
> 		k8sClient := client.InClusterClientForKarmadaApiServer()
> 		c.Set("client", k8sClient)
> 
> 		c.Next()
> 	}
> }
> ```
> 
> For example, to get deployments:
> 
> ```
> func handleGetDeployments(c *gin.Context) {
> 	if value, exists := c.Get("client"); exists {
> 		if memberClient, ok := value.(kubeclient.Interface); ok {
> 			namespace := common.ParseNamespacePathParameter(c)
> 			dataSelect := common.ParseDataSelectPathParameter(c)
> 			result, err := deployment.GetDeploymentList(memberClient, namespace, dataSelect)
> 			if err != nil {
> 				common.Fail(c, err)
> 				return
> 			}
> 			common.Success(c, result)
> 		} else {
> 			c.JSON(500, gin.H{"error": "Failed to assert memberClient to kubernetes.Interface"})
> 		}
> 	} else {
> 		c.JSON(500, gin.H{"error": "memberClient not found"})
> 	}
> }
> ```
> 
> In this case, maybe we can avoid some copy?

+1, I think it's a good practice. One more suggestion is that, maybe we can use different client name to distinguish client for k8s  and client for karmada.

Heylosky commented:
> > > I've check the code locally, the only one problem is duplicate check of membercluster. Apart from that, one question I want to talk about is that for all kind of resources in member cluster, is it need to make another copy just like the ones we implemented for the karmada apiserver. It seems that the answer is yes, but is there a better to implemented that. Just a question, i have no idea till now.
> > 
> > 
> > Perhaps add a middleware for the karmada apiserver as well, and put the client initialization process in the middleware, and then pass the client using the `c.Set` and `c.Get`? This way we can use the same implementation across all resources in the member cluster and the karmada control plane. Just like:
> > ```
> > func EnsureMemberClusterMiddleware() gin.HandlerFunc {
> > 	return func(c *gin.Context) {
> > 		karmadaClient := client.InClusterKarmadaClient()
> > 		_, err := karmadaClient.ClusterV1alpha1().Clusters().Get(context.TODO(), c.Param("clustername"), metav1.GetOptions{})
> > 		if err != nil {
> > 			c.AbortWithStatusJSON(http.StatusOK, common.BaseResponse{
> > 				Code: 500,
> > 				Msg:  err.Error(),
> > 			})
> > 			return
> > 		}
> > 		// Init client for member cluster
> > 		memberClient := client.InClusterClientForMemberCluster(c.Param("clustername"))
> > 		c.Set("client", memberClient)
> > 
> > 		c.Next()
> > 	}
> > }
> > 
> > func V1Middleware() gin.HandlerFunc {
> > 	return func(c *gin.Context) {
> > 		// Init client for controller cluster
> > 		k8sClient := client.InClusterClientForKarmadaApiServer()
> > 		c.Set("client", k8sClient)
> > 
> > 		c.Next()
> > 	}
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For example, to get deployments:
> > ```
> > func handleGetDeployments(c *gin.Context) {
> > 	if value, exists := c.Get("client"); exists {
> > 		if memberClient, ok := value.(kubeclient.Interface); ok {
> > 			namespace := common.ParseNamespacePathParameter(c)
> > 			dataSelect := common.ParseDataSelectPathParameter(c)
> > 			result, err := deployment.GetDeploymentList(memberClient, namespace, dataSelect)
> > 			if err != nil {
> > 				common.Fail(c, err)
> > 				return
> > 			}
> > 			common.Success(c, result)
> > 		} else {
> > 			c.JSON(500, gin.H{"error": "Failed to assert memberClient to kubernetes.Interface"})
> > 		}
> > 	} else {
> > 		c.JSON(500, gin.H{"error": "memberClient not found"})
> > 	}
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In this case, maybe we can avoid some copy?
> 
> +1, I think it's a good practice. One more suggestion is that, maybe we can use different client name to distinguish client for k8s and client for karmada.

You mean `c.Get("client")` this client name? But if the name in the context is different, the `func` is different also I think. (so can not reuse by member and karmada?)

warjiang commented:
> > > > I've check the code locally, the only one problem is duplicate check of membercluster. Apart from that, one question I want to talk about is that for all kind of resources in member cluster, is it need to make another copy just like the ones we implemented for the karmada apiserver. It seems that the answer is yes, but is there a better to implemented that. Just a question, i have no idea till now.
> > > 
> > > 
> > > Perhaps add a middleware for the karmada apiserver as well, and put the client initialization process in the middleware, and then pass the client using the `c.Set` and `c.Get`? This way we can use the same implementation across all resources in the member cluster and the karmada control plane. Just like:
> > > ```
> > > func EnsureMemberClusterMiddleware() gin.HandlerFunc {
> > > 	return func(c *gin.Context) {
> > > 		karmadaClient := client.InClusterKarmadaClient()
> > > 		_, err := karmadaClient.ClusterV1alpha1().Clusters().Get(context.TODO(), c.Param("clustername"), metav1.GetOptions{})
> > > 		if err != nil {
> > > 			c.AbortWithStatusJSON(http.StatusOK, common.BaseResponse{
> > > 				Code: 500,
> > > 				Msg:  err.Error(),
> > > 			})
> > > 			return
> > > 		}
> > > 		// Init client for member cluster
> > > 		memberClient := client.InClusterClientForMemberCluster(c.Param("clustername"))
> > > 		c.Set("client", memberClient)
> > > 
> > > 		c.Next()
> > > 	}
> > > }
> > > 
> > > func V1Middleware() gin.HandlerFunc {
> > > 	return func(c *gin.Context) {
> > > 		// Init client for controller cluster
> > > 		k8sClient := client.InClusterClientForKarmadaApiServer()
> > > 		c.Set("client", k8sClient)
> > > 
> > > 		c.Next()
> > > 	}
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > For example, to get deployments:
> > > ```
> > > func handleGetDeployments(c *gin.Context) {
> > > 	if value, exists := c.Get("client"); exists {
> > > 		if memberClient, ok := value.(kubeclient.Interface); ok {
> > > 			namespace := common.ParseNamespacePathParameter(c)
> > > 			dataSelect := common.ParseDataSelectPathParameter(c)
> > > 			result, err := deployment.GetDeploymentList(memberClient, namespace, dataSelect)
> > > 			if err != nil {
> > > 				common.Fail(c, err)
> > > 				return
> > > 			}
> > > 			common.Success(c, result)
> > > 		} else {
> > > 			c.JSON(500, gin.H{"error": "Failed to assert memberClient to kubernetes.Interface"})
> > > 		}
> > > 	} else {
> > > 		c.JSON(500, gin.H{"error": "memberClient not found"})
> > > 	}
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > In this case, maybe we can avoid some copy?
> > 
> > 
> > +1, I think it's a good practice. One more suggestion is that, maybe we can use different client name to distinguish client for k8s and client for karmada.
> 
> You mean `c.Get("client")` this client name? But if the name in the context is different, the `func` is different also I think. (so can not reuse by member and karmada?)

I mean for k8s client we can name the key as `kuberneters-client`, for karmada client we can name the key as `karmada-client`, for different member cluster client, we can name the key as `member-client`

Heylosky commented:
> I mean for k8s client we can name the key as `kuberneters-client`, for karmada client we can name the key as `karmada-client`, for different member cluster client, we can name the key as `member-client`

But in this case, how can we know which key should retrieve. For example, getting deployments
```
func handleGetDeployments(c *gin.Context) {
	c.Get( kuberneters-client?  or  karmada-client?  or member-client?  )
}
```
I think the route maybe like this
```
r := router.V1()
r.GET("/deployment", handleGetDeployments)
r := router.MemberV1()
r.GET("/deployment", handleGetDeployments)
```

warjiang commented:
> > I mean for k8s client we can name the key as `kuberneters-client`, for karmada client we can name the key as `karmada-client`, for different member cluster client, we can name the key as `member-client`
> 
> But in this case, how can we know which key should retrieve. For example, getting deployments
> 
> ```
> func handleGetDeployments(c *gin.Context) {
> 	c.Get( kuberneters-client?  or  karmada-client?  or member-client?  )
> }
> ```
> 
> I think the route maybe like this
> 
> ```
> r := router.V1()
> r.GET("/deployment", handleGetDeployments)
> r := router.MemberV1()
> r.GET("/deployment", handleGetDeployments)
> ```

Yeah, to take these two problems into consideration at the same time is a little bit hard, we can init the clients(for kuberneters、for karmada and for member-cluster ) in the middleware functions, and use them on demand. If we cannot reuse the resource handler currently, we can just hold on these problem, just some more duplicate code, maybe we can reduce these kinds of code in the future~

warjiang commented:
@Heylosky can you rebase your code and push again ?
![image](https://github.com/user-attachments/assets/1d0368e1-7d14-4d2b-ae5c-58d6308326d9)

Heylosky commented:
> @Heylosky can you rebase your code and push again ?

Sorry, I just push again. I have added the middleware and removed the duplicate code for the member-cluster check. I didn't add the `init clients` in the middleware functions in this PR. I think it can be add in the future code optimization for resource handler together.

warjiang commented:
/lgtm
/approve
thanks~ @Heylosky
